pkg/api/v1/clients/multicluster/multi_cluster_resource_client.go

```diff
@@ -0,0 +1,139 @@
+	}
+)
+
+type MultiClusterResourceClient interface {
+	clients.ResourceClient
+	handler.ClusterHandler
+}
+
+type multiClusterResourceClient struct {
+	resourceType    resources.Resource
+	clientGetter    ClientGetter
+	watchAggregator wrapper.WatchAggregator
+	clients         map[string]clients.ResourceClient
+	clientAccess    sync.RWMutex
+}
+
+var _ MultiClusterResourceClient = &multiClusterResourceClient{}
+
+func NewMultiClusterResourceClient(
+	resourceType resources.Resource,
+	clientGetter ClientGetter,
+	watchAggregator wrapper.WatchAggregator,
+) *multiClusterResourceClient {
+	return &multiClusterResourceClient{
+		resourceType:    resourceType,
+		clientGetter:    clientGetter,
+		watchAggregator: watchAggregator,
+		clients:         make(map[string]clients.ResourceClient),
+		clientAccess:    sync.RWMutex{},
+	}
+}
+
+func (rc *multiClusterResourceClient) Kind() string {
+	return resources.Kind(rc.resourceType)
+}
+
+func (rc *multiClusterResourceClient) NewResource() resources.Resource {
+	return resources.Clone(rc.resourceType)
+}
+
+func (rc *multiClusterResourceClient) Register() error {
+	// not implemented
+	// per-cluster clients are registered on ClusterAdded
+	return nil
+}
+
+func (rc *multiClusterResourceClient) Read(namespace, name string, opts clients.ReadOpts) (resources.Resource, error) {
+	client, err := rc.clientFor(opts.Cluster)
+	if err != nil {
+		return nil, err
+	}
+	return client.Read(namespace, name, opts)
+}
+
+func (rc *multiClusterResourceClient) Write(resource resources.Resource, opts clients.WriteOpts) (resources.Resource, error) {
+	client, err := rc.clientFor(resource.GetMetadata().Cluster)
+	if err != nil {
+		return nil, err
+	}
+	return client.Write(resource, opts)
+}
+
+func (rc *multiClusterResourceClient) Delete(namespace, name string, opts clients.DeleteOpts) error {
+	client, err := rc.clientFor(opts.Cluster)
+	if err != nil {
+		return err
+	}
+	return client.Delete(namespace, name, opts)
+}
+
+func (rc *multiClusterResourceClient) List(namespace string, opts clients.ListOpts) (resources.ResourceList, error) {
+	client, err := rc.clientFor(opts.Cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.List(namespace, opts)
+}
+
+func (rc *multiClusterResourceClient) Watch(namespace string, opts clients.WatchOpts) (<-chan resources.ResourceList, <-chan error, error) {
+	client, err := rc.clientFor(opts.Cluster)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return client.Watch(namespace, opts)
+}
+
+func (rc *multiClusterResourceClient) ClusterAdded(cluster string, restConfig *rest.Config) {
+	client, err := rc.clientGetter.GetClient(cluster, restConfig)
+	if err != nil {
+		return
+	}
+	if err := client.Register(); err != nil {
+		return
+	}
+	rc.clientAccess.Lock()
+	defer rc.clientAccess.Unlock()
+	rc.clients[cluster] = wrapper.NewClusterClient(client, cluster)
+	if rc.watchAggregator != nil {
+		rc.watchAggregator.AddWatch(wrapper.NewClusterClient(client, cluster))
```

Comment:
how is the watchAggregator used? it's not clear to me at first glance

Reply:
The multicluster resource client doesn't use the watch aggregator itself. The idea here was to allow the multicluster client to be the only place where new clients are created when clusters are added. Then users don't need to define additional cluster handlers for each resource they want to add to the aggregated watch.